macros/models/_sector/nft/enrich_trades.sql

```diff
@@ -0,0 +1,107 @@
+-- Macro to apply the NFT trades enrichment(s) to base models
+-- 1. add transaction information
+-- 2. add NFT token information
+-- 3. add ERC20 token information + handle ERC20 decimals
+-- 4. handle USD columns
+-- 5. add aggregator columns
+-- 6. fix buyer or seller for aggregator txs
+-- 7. calculate platform and royalty rates
+
+WITH base_union as (
+    {% for nft_model in models %}
+    SELECT
+        '{{ blockchain }}' as blockchain,
+        '{{ nft_model[0] }}' as project,
+        '{{ nft_model[1] }}' as project_version,
+        block_date,
+        block_number,
+        tx_hash,
+        project_contract_address,
+        trade_category,                 --buy/sell/swap
+        trade_type,                     --primary/secondary
+        buyer,
+        seller,
+        nft_contract_address,
+        nft_token_id,
+        nft_amount,                -- always 1 for erc721
+        price_raw,
+        currency_contract,
+        platform_fee_amount_raw,
+        royalty_fee_amount_raw,
+        platform_fee_address,   -- optional
+        royalty_fee_address,    -- optional
+        sub_tx_trade_id
+    FROM {{ nft_model[2] }}
+    {% if not loop.last %}
+    UNION ALL
+    {% endif %}
+    {% endfor %}
+),
+
+enrichments as (
+SELECT
+    base.blockchain,
+    base.project,
+    base.project_version,
+    base.block_date,
+    base.block_number,
+    base.tx_hash,
+    base.sub_tx_trade_id,
+    base.project_contract_address,
+    base.trade_category,
+    base.trade_type,
+    case when base.buyer != agg.contract_address then base.buyer else tx.from end as buyer,
+    case when base.seller != agg.contract_address then base.seller else tx.from end as seller,
+    base.nft_contract_address,
+    base.nft_token_id,
+    base.nft_amount,
+    base.price_raw,
+    base.currency_contract,
+    base.platform_fee_amount_raw,
+    base.royalty_fee_amount_raw,
+    base.platform_fee_address,
+    base.royalty_fee_address,
+    tx.block_time,
+    tx.from as tx_from,
+    tx.to as tx_to,
+    nft.name as nft_collection,
+    nft.standard as nft_standard,
+    erc20.symbol as currency_symbol,
+    base.price_raw/pow(10,coalesce(erc20.decimals,18)) as price,
+    base.platform_fee_amount_raw/pow(10,coalesce(erc20.decimals,18)) as platform_fee_amount,
+    base.royalty_fee_amount_raw/pow(10,coalesce(erc20.decimals,18)) as royalty_fee_amount,
+    base.price_raw/pow(10,coalesce(erc20.decimals,18))*p.price as price_usd,
+    base.platform_fee_amount_raw/pow(10,coalesce(erc20.decimals,18))*p.price as platform_fee_amount_usd,
+    base.royalty_fee_amount_raw/pow(10,coalesce(erc20.decimals,18))*p.price as royalty_fee_amount_usd,
+    agg.contract_address as aggregator_address,
+    {% if aggregator_markers != null %}
+    coalesce(agg_mark.aggregator_name, agg.name) as aggregator_name
+    {% else %}
+    agg.name as aggregator_name
+    {% endif %}
+FROM base_union base
+INNER JOIN {{ transactions_model }} tx
```

Comment:
can we make this flexible enough to filter on `block_time` as well? since the model calling this is incremental?

Reply:
I wonder if we can just include incremental blocks in the macro and if they are handled well then.

If not we can incrementally filter them before they are passed to the macro (so the `transactions_model` is then already filtered in a CTE in `nft_ethereum_trades_beta). This is not yet the case to be clear.

Reply:
:point_up: this worked